field/src/field.rs

```diff
@@ -257,6 +256,123 @@ pub trait FieldAlgebra:
+/// collection of field elements.
+///
+/// We make no guarantees about consistency of this Serialization/Deserialization
+/// across different versions of Plonky3.
+///
+/// ### Mathematical Description
+///
+/// Mathematically a more accurate name for this trait would be BasedFreeVectorSpace.
+///
+/// As `F` is a field, every field algebra `A`, over `F` is an `F`-vector space.
+/// This means we can pick a basis of elements `B = {b_0, ..., b_{n-1}}` in `A`
+/// such that, given any element `a`, we can find a unique set of `n` elements of `F`,
+/// `f_0, ..., f_{n - 1}` satisfying `a = f_0 b_0 + ... + f_{n - 1} b_{n - 1}`.
+///
+/// Thus choosing this basis `B` allows us to map between elements of `A` and
+/// arrays of `n` elements of `F`. Clearly this map depends entirely on the
+/// choice of basis `B` which may change across versions of Plonky3.
+pub trait Serializable<F: FieldAlgebra>: Sized {
+    // We could alternatively call this BasedAlgebra?
+    // The name is currently trying to indicate what this is meant to be
+    // used for as opposed to being mathematically accurate.
+
+    const DIMENSION: usize;
+
+    /// Fixes a basis for the algebra `A` and uses this to
+    /// map an element of `A` to a vector of `n` `F` elements.
+    ///
+    /// # Safety
+    ///
+    /// The value produced by this function fundamentally depends
+    /// on the choice of basis. Care must be taken
+    /// to ensure portability if these values might ever be passed to
+    /// (or rederived within) another compilation environment where a
+    /// different basis might have been used.
+    fn serialize_as_slice(&self) -> &[F];
+
+    /// Fixes a basis for the algebra `A` and uses this to
+    /// map `n` `F` elements to an element of `A`.
+    ///
+    /// # Safety
+    ///
+    /// The value produced by this function fundamentally depends
+    /// on the choice of basis. Care must be taken
+    /// to ensure portability if these values might ever be passed to
+    /// (or rederived within) another compilation environment where a
+    /// different basis might have been used.
+    #[inline]
+    fn deserialize_slice(slice: &[F]) -> Self {
+        Self::deserialize_fn(|i| slice[i].clone())
+    }
+
+    /// Fixes a basis for the algebra `A` and uses this to
+    /// map `n` `F` elements to an element of `A`. Similar
+    /// to `core:array::from_fn`, the `n` `F` elements are
+    /// given by `Fn(0), ..., Fn(n - 1)`.
+    ///
+    /// # Safety
+    ///
+    /// The value produced by this function fundamentally depends
+    /// on the choice of basis. Care must be taken
+    /// to ensure portability if these values might ever be passed to
+    /// (or rederived within) another compilation environment where a
+    /// different basis might have been used.
+    fn deserialize_fn<Fn: FnMut(usize) -> F>(f: Fn) -> Self;
+
+    /// Fixes a basis for the algebra `A` and uses this to
+    /// map `n` `F` elements to an element of `A`.
+    ///
```

Comment:
Same comment as for `deserialize_slice`, but here a sort of opposite is true too: Worth pointing out that this function does not make a fuss if the iterator contains more elements than necessary. Alternatively, perhaps it should panic (or rather, return an error) in that case?